File: code/动态规划/0005.php
<?php declare(strict_types=1);
/*
给定一个字符串 s，找到 s 中最长的回文子串。你可以假设 s 的最大长度为 1000。

示例 1：

输入: "babad"
输出: "bab"
注意: "aba" 也是一个有效答案。
示例 2：

输入: "cbbd"
输出: "bb"
*/
/**
 * 最长回文子串
 * @return [type] [description]
 *
 * 动态规划
 *
 * 时间复杂度：O(n^2)
 * 空间复杂度：O(n^2)
 */
function deal(string $str):string{
    $start_time = microtime(true);
    $start_memory = memory_get_usage(true);
    /***************代码开始*********************/
    if (strlen($str) == 0 ) return -1;
    if (strlen($str) == 1 ) return $str;
    $res = "";

    $len = strlen($str);
    $dp = array_fill(0, $len+2, array_fill(0, $len+2, 0));
    for ($i=1; $i <= $len; $i++) {
        for ($j=0; $j < $len; $j++) {
            $end = $j + $i - 1;
            if ($end >= $len) {
                break;
            }
            $dp[$j][$end] = ($i == 1 || $i == 2 || $dp[$j+1][$end-1]) && $str[$j] == $str[$end];
            if($dp[$j][$end] && $i > strlen($res)){
                $res = substr($str, $j, $i);
            }
        }
    }


    /***************代码结束*******************/
    /***********性能监控***********************/
    $end_time = microtime(true);
    $need_time = $end_time - $start_time;
    $end_memory = memory_get_usage(true);
    $memory_peak = memory_get_peak_usage(true);
    print_r("耗时:" . $need_time . "\r\n");
    print_r("开始内存:" . $start_memory . "\r\n");
    print_r("结束内存:" . $end_memory . "\r\n");
    print_r("峰值内存:" . $memory_peak . "\r\n");
    /*******************************************/
    return $res;
}
echo(deal("babad"));
echo "\r\n";
echo(deal("cbbd"));

/**
 * 中心扩展法
 * 时间复杂度：O(n^2)
 * 空间复杂度：O(1)
 */
function deal2(string $str):string{
    $start_time = microtime(true);
    $start_memory = memory_get_usage(true);
    /***************代码开始*********************/
    $res = "";
    $len = strlen($str);
    $end = 2 * $len - 1;
    for ($i=0; $i < $end; $i++) {
        $mid = $i / 2;
        $p = (int)floor($mid);
        $q = (int)ceil($mid);

        while ($p>=0 && $q < $len) {
            if($str[$p] != $str[$q]) break;
            $p--;
            $q++;
        }
        $l = $q- $p - 1;
        if($l > strlen($res))
            $res = substr($str, $p+1, $l);
    }



    /***************代码结束*******************/
    /***********性能监控***********************/
    $end_time = microtime(true);
    $need_time = $end_time - $start_time;
    $end_memory = memory_get_usage(true);
    $memory_peak = memory_get_peak_usage(true);
    print_r("耗时:" . $need_time . "\r\n");
    print_r("开始内存:" . $start_memory . "\r\n");
    print_r("结束内存:" . $end_memory . "\r\n");
    print_r("峰值内存:" . $memory_peak . "\r\n");
    /*******************************************/
    return $res;
}
echo(deal2("babad"));
echo "\r\n";
echo(deal2("cbbd"));

/**
 * 马拉车算法  Manacher
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */
function deal3(string $str):string{
    $start_time = microtime(true);
    $start_memory = memory_get_usage(true);
    /***************代码开始*********************/
    $res = "";
    $len = strlen($str);

    if ($str == "") return "";
    $len = strlen($str);
    if($len == 1 || $str == strrev($str)) return $str;

    $mlen = 1;
    $start = 0;
    for ($i=1; $i < $len; $i++) {
        $even = substr($str, $i-$mlen, $mlen+1);
        $odd = substr($str, $i-$mlen-1, $mlen+2);

        if($i - $mlen - 1 >= 0 && $odd == strrev($odd)){
            $start = $i - $mlen - 1;
            $mlen += 2;
            continue;
        }
        if($i-$mlen >= 0 && $even == strrev($even)){
            $start = $i - $mlen;
            $mlen += 1;
            continue;
        }
    }
    $res = substr($str, $start, $mlen);




    /***************代码结束*******************/
    /***********性能监控***********************/
    $end_time = microtime(true);
    $need_time = $end_time - $start_time;
    $end_memory = memory_get_usage(true);
    $memory_peak = memory_get_peak_usage(true);
    print_r("耗时:" . $need_time . "\r\n");
    print_r("开始内存:" . $start_memory . "\r\n");
    print_r("结束内存:" . $end_memory . "\r\n");
    print_r("峰值内存:" . $memory_peak . "\r\n");
    /*******************************************/
    return $res;
}
echo(deal3("babad"));
echo "\r\n";
echo(deal3("cbbd"));

/**
 * 思路是，因为题目要的最长回文，所以不用验证更短的子串了。

或者说，就是以尾座标为标志进行验证，也不用担心错过更长的结果。效果是只需遍历一次。时间空间都是双百
 */
function deal4(string $str):string{
    $start_time = microtime(true);
    $start_memory = memory_get_usage(true);
    /***************代码开始*********************/
    $res = "";
    $len = strlen($str);

    if ($str == "") return "";
    $len = strlen($str);
    if($len == 1 || $str == strrev($str)) return $str;

    $mlen = 1;
    $start = 0;
    for ($i=1; $i < $len; $i++) {
        $even = substr($str, $i-$mlen, $mlen+1);
        $odd = substr($str, $i-$mlen-1, $mlen+2);

        if($i - $mlen - 1 >= 0 && $odd == strrev($odd)){
            $start = $i - $mlen - 1;
            $mlen += 2;
            continue;
        }
        if($i-$mlen >= 0 && $even == strrev($even)){
            $start = $i - $mlen;
            $mlen += 1;
            continue;
        }
    }
    $res = substr($str, $start, $mlen);




    /***************代码结束*******************/
    /***********性能监控***********************/
    $end_time = microtime(true);
    $need_time = $end_time - $start_time;
    $end_memory = memory_get_usage(true);
    $memory_peak = memory_get_peak_usage(true);
    print_r("耗时:" . $need_time . "\r\n");
    print_r("开始内存:" . $start_memory . "\r\n");
    print_r("结束内存:" . $end_memory . "\r\n");
    print_r("峰值内存:" . $memory_peak . "\r\n");
    /*******************************************/
    return $res;
}
echo(deal4("babad"));
echo "\r\n";
echo(deal4("cbbd"));

File: code/动态规划/0005.py
# 解题思路

# 思路是，因为题目要的最长回文，所以不用验证更短的子串了。

# 或者说，就是以尾座标为标志进行验证，也不用担心错过更长的结果。效果是只需遍历一次。时间空间都是双百。


def deal(s: str) -> str:
    if not s:
        return ""
    length = len(s)
    if length == 1 or s == s[::-1]:
        return s
    mlen, start = 1, 0

    for i in range(1, length):
        even = s[i-mlen:i+1]
        odd = s[i-mlen-1:i+1]

        if i-mlen - 1 > 0 and odd == odd[::-1]:
            start = i-mlen-1
            mlen += 2
            continue
        if i-mlen >= 0 and even == even[::-1]:
            start = i-mlen
            mlen += 1
            continue
    return s[start:start+mlen]


print(deal("babad"))


def deal2(s: str) -> str:
    right, left, count = 0, 0, 1
    start = 0
    mlen = 0

    slen = len(s)

    for i in range(0, slen, count):
        count = 1
        left = i -1
        right = i + 1
        #选出重复字符
        while(right<slen and s[i] == s[right]): # 减少了大量操作
            right += 1
            print("***"+s[i]+"\r\n")
            print(right)
            count += 1
        #由中心字符向左右扩展
        while(left>=0 and right<slen and s[left]==s[right]):
            right += 1
            left -= 1

        if mlen < (right - left - 1): # 取最长
            #左右标记不在回文子串范围内，在外面两侧，需要减1
            mlen = right - left - 1
            start = left + 1
    return s[start:mlen]



print(deal2("babad"))
print(deal2("abbaba"))



'''
最详细解释和思路
char * longestPalindrome(char * s){
    //判断字符串长度，处理特殊情况，字符数组为空或只有1个字符时('\0'结束符不占长度)，本身就是题解
    if (strlen(s) == 0 || strlen(s) == 1) {
        return s;
    }

    //对于>=2个字符以上时，则对每个字符进行循环查找，无非分为/1、回文子串字符数为偶数则从i向右查找
    //2、若回文子串字符数为奇数个，则从i-1和i+1位置比较
    //另外，注意第1种情况偶数字符数的回文子串可以当做1个字符考虑，这样节省寻找时间。
    int i, left, right, count = 0; //注意初始化时若不显式初始化为0相当于未初始化，使用时必须先初始化
    int len = 0; //最后回文子串的长度
    int start = 0; //最后回文子串的起始位置
    for (i = 0; s[i] != '\0'; i += count) {
        count = 1; //每次for循环进来都初始为1，重新判断i字符右面有无重复字符，若有则下次i+count略过重复字符
        left = i - 1;
        right = i + 1;

        //右面字符非结束符前提下，比较当前字符和右面字符是否相等---先处理字符重复的情况
        while (s[right] != '\0' && s[i] == s[right]) {
            right++;
            count++; //若相等则count+1，下次略过
        }

        //左右字符合法前提下，比较左面和右面字符是否相等---处理所有场景下回文字符的情况
        while (left >= 0 && s[right] != '\0' && s[left] == s[right]) {
            left--;
            right++;
        }

        //如果当前回文子串长度大于上一次长度，则刷新起始位置和长度
        if (right - left - 1 > len) {
            len = right - left - 1; //注意真正的回文子串在right和left中间这一段，所以 -1
            start = left + 1;
        }
    }

    s[start + len] = '\0'; //最后将长度后的一个字符赋值为结束符
    return s + start; //返回字符数组即指针+偏移
}

作者：jet534
链接：https://leetcode-cn.com/problems/longest-palindromic-substring/solution/zui-xiang-xi-jie-shi-he-si-lu-by-jet534/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

'''



